close of closed channel in router.go

Encountered this error while debugging my irc_output plugin's restartability. Essentially this error only happened after successfully restarting my plugin through the Restarting interface.

```
panic: runtime error: close of closed channel [recovered]
    panic: runtime error: close of closed channel

goroutine 89 [running]:
runtime.panic(0x4503ec0, 0x482b155)
    /usr/local/go/src/pkg/runtime/panic.c:279 +0xf5
github.com/mozilla-services/heka/pipeline.func·018()
    /Users/chance/projects/heka/build/heka/src/github.com/mozilla-services/heka/pipeline/router.go:241 +0x10d
runtime.panic(0x4503ec0, 0x482b155)
    /usr/local/go/src/pkg/runtime/panic.c:248 +0x18d
github.com/mozilla-services/heka/pipeline.func·019()
    /Users/chance/projects/heka/build/heka/src/github.com/mozilla-services/heka/pipeline/router.go:294 +0x39a
created by github.com/mozilla-services/heka/pipeline.(*MatchRunner).Start
    /Users/chance/projects/heka/build/heka/src/github.com/mozilla-services/heka/pipeline/router.go:295 +0xc8
```

--- a/CHANGES.txt
+++ b/CHANGES.txt
@@ -24,6 +24,9 @@ Backwards Incompatibilities
 Bug Handling
 ------------
 
+* Prevent panics during shutdown when a restarting plugin has been restarted
+  (issue #957).
+
 * Added support for a Logstreamer symbolically linked log_directory root 
   (issue #741).
 

--- a/pipeline/plugin_runners.go
+++ b/pipeline/plugin_runners.go
@@ -490,12 +490,12 @@ func (foRunner *foRunner) Starter(h PluginHelper, wg *sync.WaitGroup) {
 
 	var pw *PluginWrapper
 
-	for !globals.Stopping {
-		if foRunner.matcher != nil {
-			sampleDenom := h.PipelineConfig().Globals.SampleDenominator
-			foRunner.matcher.Start(foRunner.inChan, sampleDenom)
-		}
+	if foRunner.matcher != nil {
+		sampleDenom := globals.SampleDenominator
+		foRunner.matcher.Start(foRunner.inChan, sampleDenom)
+	}
 
+	for !globals.Stopping {
 		// `Run` method only returns if there's an error or we're shutting
 		// down.
 		if filter, ok := foRunner.plugin.(Filter); ok {
@@ -515,7 +515,7 @@ func (foRunner *foRunner) Starter(h PluginHelper, wg *sync.WaitGroup) {
 			foRunner.LogMessage("stopped")
 		}
 
-		// Are we supposed to stop? Save ourselves some time by exiting now
+		// Are we supposed to stop? Save ourselves some time by exiting now.
 		if globals.Stopping {
 			return
 		}
@@ -527,10 +527,10 @@ func (foRunner *foRunner) Starter(h PluginHelper, wg *sync.WaitGroup) {
 		}
 
 		if pw == nil {
-			return // no wrapper means it is Stoppable
+			return // no wrapper means it is Stoppable.
 		}
 
-		// We stop and let this quit if its not a restarting plugin
+		// We stop and let this quit if its not a restarting plugin.
 		if recon, ok := foRunner.plugin.(Restarting); ok {
 			recon.CleanupForRestart()
 		} else {
@@ -539,9 +539,8 @@ func (foRunner *foRunner) Starter(h PluginHelper, wg *sync.WaitGroup) {
 			return
 		}
 
-		// Re-initialize our plugin using its wrapper
-		// Attempt to recreate the plugin until it works without error
-		// or until we were told to stop
+		// Re-initialize our plugin using its wrapper. Attempt to recreate the
+		// plugin until it works without error or until we're told to stop.
 	createLoop:
 		for !globals.Stopping {
 			err = rh.Wait()

